qt/scientific_interfaces/Indirect/IndirectFunctionBrowser/MSDFunctionModel.cpp

```diff
@@ -0,0 +1,520 @@
+    } else {
+      throw std::runtime_error("Cannot set function " + name);
+    }
+    m_model.setFunction(fun);
+    return;
+  }
+  bool isFitTypeSet = false;
+  for (size_t i = 0; i < fun->nFunctions(); ++i) {
+    const auto f = fun->getFunction(i);
+    const auto name = f->name();
+    if (name == "MsdGauss" || name == "MsdPeters" || name == "MsdYi") {
+      if (isFitTypeSet) {
+        throw std::runtime_error("Function has wrong structure.");
+      }
+      m_fitType = QString::fromStdString(name);
+      isFitTypeSet = true;
+    } else {
+      clear();
+      throw std::runtime_error("Function has wrong structure.");
+    }
+  }
+  m_model.setFunction(fun);
+}
+
+IFunction_sptr MSDFunctionModel::getFitFunction() const {
+  return m_model.getFitFunction();
+}
+
+bool MSDFunctionModel::hasFunction() const { return m_model.hasFunction(); }
+
+void MSDFunctionModel::addFunction(const QString &prefix,
+                                   const QString &funStr) {
+  if (!prefix.isEmpty())
+    throw std::runtime_error(
+        "Function doesn't have member function with prefix " +
+        prefix.toStdString());
+  const auto fun =
+      FunctionFactory::Instance().createInitialized(funStr.toStdString());
+  const auto name = fun->name();
+  QString newPrefix;
+  if (name == "MsdGauss" || name == "MsdPeters" || name == "MsdYi") {
+    setFitType(QString::fromStdString(name));
+    newPrefix = *getFitTypePrefix();
+  } else {
+    throw std::runtime_error("Cannot add function " + name);
+  }
+  auto newFun = getFunctionWithPrefix(newPrefix, getSingleFunction(0));
+  copyParametersAndErrors(*fun, *newFun);
+  if (getNumberLocalFunctions() > 1) {
+    copyParametersAndErrorsToAllLocalFunctions(*getSingleFunction(0));
+  }
+}
+
+void MSDFunctionModel::removeFunction(const QString &prefix) {
+  if (prefix.isEmpty()) {
+    clear();
+    return;
+  }
+  auto prefix1 = getFitTypePrefix();
+  if (prefix1 && *prefix1 == prefix) {
+    removeFitType();
+    return;
+  }
+  throw std::runtime_error(
+      "Function doesn't have member function with prefix " +
+      prefix.toStdString());
+}
+
+void MSDFunctionModel::setFitType(const QString &name) {
+  const auto oldValues = getCurrentValues();
+  m_fitType = name;
+  m_model.setFunctionString(buildFunctionString());
+  m_model.setGlobalParameters(makeGlobalList());
+  setCurrentValues(oldValues);
+}
+
+void MSDFunctionModel::removeFitType() {
+  const auto oldValues = getCurrentValues();
+  m_fitType.clear();
+  m_model.setFunctionString(buildFunctionString());
+  m_model.setGlobalParameters(makeGlobalList());
+  setCurrentValues(oldValues);
+}
+
+bool MSDFunctionModel::hasGaussianType() const {
+  return m_fitType == "MsdGauss";
+}
+
+bool MSDFunctionModel::hasPetersType() const {
+  return m_fitType == "MsdPeters";
+}
+
+bool MSDFunctionModel::hasYiType() const { return m_fitType == "MsdYi"; }
+
+void MSDFunctionModel::updateParameterEstimationData(
```

Comment:
Every other method in this file uses `setX` or `setY` whilst this is a setter too which uses `updateX`, could we change it to `setParameterEstimationData`